Allow overriding the locale for individual translations

grantovich opened the pull request:
I had some free time, so here is my initial stab at #298. I couldn't find a way to stub the ENV value so I could test that it's correctly loaded into the service, but I manually confirmed that it works.

grantovich commented on addon/services/i18n.js:
```diff
@@ -19,11 +19,23 @@ export default Parent.extend(Ember.Evented, {
+      if (this.get('allowLocaleOverride') === undefined) {
+        Ember.warn('Specifying an interpolation key named `locale` now overrides the target locale for translation. To silence this warning, set ENV.i18n.allowLocaleOverride to true. To disable this behavior and treat `locale` as a normal interpolation key, set ENV.i18n.allowLocaleOverride to false.');
+      }
+      locale = new Locale(data.locale, this.container);
```
This part seems inefficient, but addressing it would probably require a larger change (for instance, locales could be cached on the service when they are initialized, or all locales could be loaded up-front).

grantovich commented on addon/services/i18n.js:
```diff
@@ -19,11 +19,26 @@ export default Parent.extend(Ember.Evented, {
+          { id: 'ember-i18n.locale-override' }
+        );
+      }
+      locale = new Locale(data.locale, this.container);
```
This part seems inefficient, but addressing it would probably require a larger change (for instance, locales could be cached on the service when they are initialized).

grantovich commented on addon/services/i18n.js:
```diff
@@ -19,11 +19,25 @@ export default Parent.extend(Ember.Evented, {
+        (this.get('allowLocaleOverride') !== undefined),
+        { id: 'ember-i18n.locale-override' }
+      );
+      locale = new Locale(data.locale, this.container);
```
This part seems inefficient, but addressing it would probably require a larger change (for instance, locales could be cached on the service when they are initialized).

jamesarosen replied:
1. we only want to warn if `data.locale` is defined
 2. we can make it slightly more efficient by saving off the warning text as a `const` at the top of the file
 3. we could cache `Locale`s in a closure-local map in the `locale.js` file:

```js
const seenLocales = {};

function Locale(name, container) {
  if (seenLocales[name]) { return seenLocales[name]; }
  seenLocales[name] = this;
  ...
}
```

grantovich replied:
> we only want to warn if `data.locale` is defined

This is inside the `if (data.locale && ...)` so the warning should only be seen in that case.

> we could cache `Locale`s in a closure-local map in the `locale.js` file

Clever! I will see if this works.

jamesarosen replied:
> This is inside the if (data.locale && ...)

d'oh

grantovich replied:
Though I can't help thinking that from the outside, it's very counter-intuitive that `new Locale` only *really* constructs a new locale the first time it's called with a given argument, and thereafter just returns the previously-constructed object. At that point it's really more like a lazily-populated hash. Maybe a "locales" service or something?

jamesarosen replied:
That file is private, so we can change its API. We can change the exported function to `localeFor(name, container)` and uses the map.

grantovich replied:
Sounds like a good simple solution, I will try going in that direction.

jamesarosen commented on app/instance-initializers/ember-i18n.js:
```diff
@@ -14,6 +14,7 @@ export default {
     const key = 'service:i18n';
     const i18n = instance.lookup ? instance.lookup(key) : instance.container.lookup(key);
     i18n.set('locale', defaultLocale);
+    i18n.set('allowLocaleOverride', (ENV.i18n || {}).allowLocaleOverride);
```
Because this initializer doesn't run in all situations (notably non-integration tests), I suggest moving this logic into the Service itself:
```js
allowLocaleOverride: Ember.computed({
  get() { return (ENV.i18n || {}).allowLocaleOverride; },
  set(prop, newValue) { return newValue; }
})
```

grantovich replied:
I think it's probably a good idea that all the code which initializes the service with ENV values should be in the same place – it would be strange to have one property over here, and one property over there. Or would you also suggest moving the `defaultLocale` stuff? (that seems like a separate change)

jamesarosen replied:
Yes, I think that makes sense as well. I'll do that now and link the PR.

jamesarosen replied:
Except the service is in `addon/` and the config is in `app/`. Hmm. I don't know that there's a way to reference `ENV` from the service directly :(

grantovich replied:
I think it's fine – this way the service itself is isolated, and doesn't try to pull in outside information when you unit-test it. Presumably the initializer logic could be covered by an integration test (if there was a way to stub ENV values).

jamesarosen replied:
I meant that in an app's unit tests, they're not going to get the same behavior that they get in an integration test. It's going to fail in a way that's hard to diagnose.

grantovich replied:
Ah, I see, you were talking about the *user's* integration tests and unit tests. My bad.

jamesarosen replied:
From elwayman02/ember-data-echonest:

```
const ENV = this.container.lookupFactory('config:environment');
```

jamesarosen replied:
You may want to rebase off of master once #313 is merged.

jamesarosen commented on tests/unit/i18n-t-test.js:
```diff
@@ -71,3 +88,27 @@ test("check unknown locale", function(assert) {
+  const result = i18n.t('with.locale.interpolation', { locale: 'es' });
+  assert.equal(result.string, 'Locale: es');
+  assert.equal(warnings.length, 0);
+});
```
These tests are awesome :)

grantovich commented on tests/unit/i18n-t-test.js:
```diff
@@ -71,3 +88,27 @@ test("check unknown locale", function(assert) {
+  const i18n = this.subject({ locale: 'en', allowLocaleOverride: false });
+
+  const result = i18n.t('with.locale.interpolation', { locale: 'es' });
+  assert.equal(result.string, 'Locale: es');
```
For some reason, in these tests alone, the `t` function seems to return an object with a single property called `'string'`, which I have to access to get the actual translation. Really not sure why that is.

grantovich commented:
@jamesarosen I tried changing things to use your `localeFor`-style proposal, but that caused test failures related to adding translations at runtime. Strangely, using your originally proposed approach (returning from the constructor) does work.

I've addressed the outstanding feedback and rebased. The `.string` issue seems to have mysteriously disappeared... so I guess this should be good to go.

grantovich commented:
@jamesarosen Any further changes needed? Could this be merged and released? (I guess it should have a changelog entry, I could write that if you like)

grantovich commented:
When testing this in my app, I discovered a need to find whether a given translation exists in a locale other than the current one. So I extracted `_getLocaleWithOverride` and applied it to `exists` as well. Not sure what's up with the test failures on Travis – I don't get them locally, and they seem unrelated to my changes.

jamesarosen commented on addon/utils/locale.js:
```diff
@@ -13,6 +15,10 @@ export default class Locale {
   constructor(id, container) {
+    if (initializedLocales[id]) {
+      return initializedLocales[id];
+    }
```
You never populate `initializedLocales` with `this` after the `if`.

grantovich replied:
Argh. You're completely right. Unfortunately, actually doing this causes adding translations at runtime to stop working. So I guess this method of caching won't work after all. I'm really not sure how to handle this correctly, the architecture is hard for me to follow... should I just take it out for now? I can personally live with the sub-optimal performance, since my use case is not very intensive (this may be a case of premature optimization).

jamesarosen replied:
Can you try removing the `if` around `locale.rebuild()` in `addTranslations`?

I think we need to do some substantial re-architecting anyway to support #255. Probably best to think about the two problems together.

grantovich replied:
Just tried that, but the test for adding translations at runtime still fails.

jamesarosen replied:
OK. I'll give #255 a shot first. I may try moving some of the `addTranslations` logic into `Locale` so the `I18nService` doesn't have to worry about state management.

grantovich replied:
Alright. For now I'll proceed with using this branch in my app, since it seems stable and I haven't noticed any performance hits (I have a small number of translations, and the cross-locale lookup is only performed once on page load). I will keep an eye on that task.

jamesarosen commented on addon/services/i18n.js:
```diff
@@ -41,7 +46,7 @@ export default Parent.extend(Evented, {
   // @public
   exists: function(key, data = {}) {
-    const locale = this.get('_locale');
+    const locale = this._getLocaleWithOverride(data);
```
I like the symmetry of `t` and `exists` now. I'm glad you did that extraction.

grantovich commented:
@jamesarosen I've resolved conflicts with current master and removed the defunct attempt at caching. How would you like to handle this? Should it still wait on #255? Can it be merged with a potential caveat about performance? (I've been using this in production for a month now and haven't seen any breaking issues with it, but also haven't examined performance too closely)

jamesarosen commented:
Sorry I've let this sit so long. I want to work on #255 very soon. I'll try to eke out some time this week. Once I have a better picture of that work, I'll let you know what we should do with this. I definitely want either this code or this idea in the project.

grantovich commented:
Rebased this on v4.2.0 to resolve conflicts. @jamesarosen how's it going? :smiley:

jamesarosen commented:
I'm hoping to do some of this heavier work over the break right before the new year. I'll keep you updated.

grantovich commented:
Hi @jamesarosen, just want to confirm whether this is still slated to be merged at some point. In our limited use case we've had no problems with it, and have been using it in production since October – but we don't want to maintain a fork of ember-i18n indefinitely.

Of course, I'd be happy to write a changelog entry and wiki documentation for this feature, or any other friction-reducing tasks that would help.

jamesarosen commented:
There's not much movement on ember-i18n these days because I'm working on a release for another project right now. But after that release will be some I18n work, which will be a perfect time to pick this up again.

Thanks for testing it out!

sbsurf commented:
This would definitely be useful. I could use this as a way to inject missing translation fallbacks, while overriding the missingMessage function to just return the english translation.

sandstrom commented:
@jamesarosen This would be a wonderful addition! ⛵ 

Is there anything in particular holding this PR back?

sandstrom commented:
@grantovich Great initiative! 

What do you think about this line:
https://github.com/jamesarosen/ember-i18n/blob/9e840285b53879e5a1eeb7fca57076b69b6f8c9b/addon/services/i18n.js#L41

```js
this.trigger('missing', this.get('locale'), key, data); # current
this.trigger('missing', locale.id, key, data); # my suggestion
```

grantovich commented:
@sandstrom Thanks for the catch – I haven't been keeping this up-to-date lately since it seems there isn't much interest in merging it, and our project that uses this fork is in maintenance mode. If @jamesarosen gives the go-ahead, I would be happy to revisit it.

(tagging @jasonmit since from https://github.com/jamesarosen/ember-i18n/pull/381#issuecomment-230182277 it looks like we might be duplicating some effort)

jpsirois commented:
I just had this exact same need for a project. And I extended the i18n services with the changes from this PR and it just work. Is this something still in consideration? I think this PR is worth consideration.

sandstrom commented:
friendly ping @jamesarosen 😄

jamesarosen commented:
I don't have any objections. I'd love one of the other maintainers to take a look and help shepherd this.

jamesarosen commented:
jamesarosen/ember-i18n has been deprecated in favor of [ember-intl](https://github.com/ember-intl).
